src/entities/nestedJoin/index.ts

```diff
@@ -0,0 +1,270 @@
+import {
+    createPublicClient,
+    decodeAbiParameters,
+    decodeFunctionResult,
+    encodeFunctionData,
+    http,
+} from 'viem';
+import { Address, Hex } from '../../types';
+import { Slippage } from '../slippage';
+import { Token } from '../token';
+import { BALANCER_RELAYER, CHAINS, getPoolAddress } from '../../utils';
+import { Relayer } from '../relayer';
+import { parseNestedJoinArgs } from '../utils/parseNestedJoinArgs';
+import { TokenAmount } from '../tokenAmount';
+import { balancerRelayerAbi, bathcRelayerLibraryAbi } from '../../abi';
+
+export type NestedJoinInput = {
+    amountsIn: {
+        address: Address;
+        rawAmount: bigint;
+    }[];
+    chainId: number;
+    rpcUrl: string;
+    testAddress: Address;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+    fromInternalBalance?: boolean;
+};
+
+export type JoinStep = {
+    action: 'join'; // TODO: are we supporting other actions, such as swaps?
+    level: number; // 0 is the bottom and the highest level is the top
+    input: NestedToken;
+    poolId: Address; // output poolId
+    isTop: boolean;
+};
+
+export type NestedPoolState = {
+    pools: {
+        id: Hex;
+        address: Address;
+        type: string;
+    }[];
+    joinSteps: JoinStep[]; // each token should have at least one
+};
+
+type NestedToken = {
+    address: Address;
+    decimals: number;
+    index: number;
+};
+
+export type NestedJoinArgs = {
+    chainId: number;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+    sortedTokens: Token[];
+    poolId: Address;
+    poolType: string;
+    kind: number;
+    sender: Address;
+    recipient: Address;
+    maxAmountsIn: {
+        amount: bigint;
+        isRef: boolean;
+    }[];
+    minBptOut: bigint;
+    fromInternalBalance: boolean;
+    outputReferenceKey: bigint;
+};
+
+type NestedJoinQueryResult = {
+    calls: NestedJoinArgs[];
+    bptOut: TokenAmount;
+};
+
+type NestedJoinCallInput = NestedJoinQueryResult & {
+    slippage: Slippage;
+    sender: Address;
+    recipient: Address;
+    relayerApprovalSignature?: Hex;
+};
+
+const outputRefOffset = 100n;
+
+export class NestedJoin {
+    async query(
+        input: NestedJoinInput,
+        nestedPoolState: NestedPoolState,
+    ): Promise<NestedJoinQueryResult> {
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        // sort pools by ascending level
+        // then go from bottom pool to up filling out input and output amounts from joinSteps
+        // input at level 0 are either amountsIn provided or 0n
+        // input at following levels can be amountsIn provided, output of the previous level or 0n
+        // output at max level is the bptOut
+
+        const stepsSortedByLevel = nestedPoolState.joinSteps.sort(
```

Comment:
Feels like `NestedPoolState` data structure could be changed to be more useful as we're really just manipulating it before its being used. Might be better to have each pool contain its joinSteps and level? something like:
```
    pools: {
        id: Hex;
        address: Address;
        type: string;
        level: number;
        joinSteps: JoinStep[]; // each token should have at least one
    }[];
  
```

Reply:
Yeap, I agree it would be nice to move data to inside the pools. 👍

Reply:
And thinking a bit more about the data structure, since `level` was pulled out to the pool and `isTop` is not actually being used at the moment, I'd say we could move to something like this:
```
    pools: {
        id: Hex;
        address: Address;
        type: string;
        level: number;
        nestedTokens: {
            address: Address;
            balance: bigint;
            decimals: number;
            index: number;
            action: 'join';
        }[];
    }[];
```

Reply:
If you rename `nestedTokens` to just `tokens` it keeps consistency and ends up being basically the same as the normal join but with a bit of extra data.

Reply:
Ok 👍  We could even remove the action 'join' while we don't have any other actions. And if/when we add them in the future, they could be a pool rather than a token data.